assignments/nested-parens.adoc

```diff
@@ -0,0 +1,228 @@
+= Exercise: Nested Parens
+:source-language: rust
+
+In this exercise, you will learn
+
+* how to translate a `for loop` into an `iterator` solution by using `map`, `scan`, and `fold`
+* how to setup a benchmarking harness to compare speeds
+* how to use `Rayon` to parallelize your iterators!
+* how to compress data to get even more speedups
+
+We will take a famous coding problem and iterate solutions, and measure their performance.
+Problem statement: Given a well-formed set of `String` parentheses (one entry per line, all lines in a single text file), 
+
+[source, text, linenums]
+----
+()(())
+(())()
+(()())
+()()()
+((()))
+
+----
+
+find the line where there is the maximum number of nested parentheses. 
+In the above example, it would be one line `5`, as the maximum nesting level is `3`.
+
+
+In this assignment, you will be given:
+* Code to write to a local file the parenthesis inputs (which you can assume are only nested correctly)
+* A working solution via a `for loop`, and it will be your task to translate it to an iterator style, and then to benchmark the difference.
+
+
+=== Step 1
+1. Make a new cargo crate with
+
+[source, bash]
+----
+cargo new nested
+----
+
+and copy the following code in your `src/main.rs`:
+2. Complete some of the missing blanks of the code, and make sure `cargo test` passes.
+
+[source,rust, linenums]
+====
+----
+use std::env;
+use std::fs::File;
+use std::io::prelude::*;
+
+// Adapted from https://haggainuchi.com/nestedparens.html
+fn nested(n: i32) -> Vec<String> {
+    if n == 0 {
+        // 👀 How did I make an empty `String` again?
+        vec![XXX]
```

Comment:
Maybe
```suggestion
        vec![todo!("👀 How do I make an empty `String` again?")]
```
or
```suggestion        
        // TODO: 👀 How did I make an empty `String` again?
        vec![XXX]
```

Reply:
Or maybe I'd state somewhere before the code to look for :eyes: (see what I did there?) and complete the code there